### PR TITLE
Add Numerical Sequence Stype

### PR DESCRIPTION
**Pull request by yiweny**
Sequence stype is very similar to multi_categorical stype, so I might as well add it.

### CODE REVIEW

File: torch_frame/data/stats.py [hunk 1]
```diff
@@ -62,16 +54,15 @@ def compute(self, ser: Series, sep: Optional[str] = None) -> Any:
...
         elif self == StatType.MULTI_COUNT:
             assert sep is not None
-            print("sep is ", sep)
             ser = ser.apply(
                 lambda x: set([cat.strip() for cat in x.split(sep)])
                 if (x is not None and x != '') else set())
             ser = ser.explode().dropna()
             count = ser.value_counts(ascending=False)
             return count.index.tolist(), count.values.tolist()
-        raise NotImplementedError(f"Invalid stat type '{self.value}'")
```
Comment (yiweny) [inline torch_frame/data/stats.py:74]: Don't think it's possible to hit this error

File: torch_frame/data/stats.py [hunk 2]
```diff
@@ -30,24 +30,16 @@ class StatType(Enum):
...
-                StatType.COUNT,
-            ]
-        elif stype == torch_frame.text_embedded:
-            return []
-        elif stype == torch_frame.multicategorical:
-            return [
-                StatType.MULTI_COUNT,
-            ]
-
-        raise NotImplementedError(f"Invalid semantic type '{stype.value}'")
```
Comment (yiweny) [inline torch_frame/data/stats.py:50]: I don't think it's possible to hit this error.

**Comment by weihua916** [comment]
For sequence stype, I think we should directly operate on a Python list, rather than parsing as we do for multi-categorical columns:
```python
>>> pd.DataFrame({'seq': [[0, 1, 2], [3, 4]]})
         seq
0  [0, 1, 2]
1     [3, 4]
>>>
```
Also, I think we should distinguish numerical sequence from the categorical sequence. Let's discuss more.

**Comment by rusty1s** [comment]
> For sequence stype, I think we should directly operate on a Python list, rather than parsing as we do for multi-categorical columns.

+1

File: test/data/test_mapper.py [hunk 3]
```diff
@@ -65,6 +67,26 @@ def test_multicategorical_tensor_mapper():
     assert out.values[5] == ''
 
 
+def test_sequence_tensor_mapper():
+    ser = pd.Series([[0.1, 0.5], [0.3], [], [0.2, np.nan], None, np.nan])
+    expected_values = torch.tensor([0.1, 0.5, 0.3, 0.2, torch.nan],
+                                   dtype=torch.float64)
```
Comment (weihua916) [inline test/data/test_mapper.py:73]: Can we use/cast to `float32`?

File: test/data/test_mapper.py [hunk 4]
```diff
@@ -65,6 +67,26 @@ def test_multicategorical_tensor_mapper():
...
+    expected_boundaries = torch.tensor([0, 2, 3, 3, 5, 5, 5])
```
Comment (weihua916) [inline test/data/test_mapper.py:74]: nit: expected_offset

File: test/data/test_dataset.py [hunk 5]
```diff
@@ -88,9 +88,10 @@ def test_dataset_inductive_transform():
 
 def test_converter():
     dataset = FakeDataset(
-        num_rows=10,
-        stypes=[stype.categorical, stype.numerical,
-                stype.multicategorical]).materialize()
+        num_rows=10, stypes=[
+            stype.categorical, stype.numerical, stype.multicategorical,
+            stype.sequence
```
Comment (weihua916) [inline test/data/test_dataset.py:93]: Should we rename it as `stype.sequence_numerical`?

File: torch_frame/data/dataset.py [hunk 6]
```diff
@@ -211,10 +215,10 @@ class Dataset(ABC):
...
             :obj:`2`. (default: :obj:`None`).
         col_to_sep (Union[str, Dict[str, str]]): A dictionary or a string
-            specifying the separator/delimiter for the multi-categorical
-            columns. If a string is specified, then the same separator will be
-            used throughout all the multi-categorical columns. If a dictionary
-            is given, we use a separator specified for each column.
+            specifying the separator/delimiter for the multi-categorical or
+            sequence columns. If a string is specified, then the same separator
+            will be used throughout all the multi-categorical columns. If a
+            dictionary is given, we use a separator specified for each column.
```
Comment (weihua916) [inline torch_frame/data/dataset.py:221]: Revert this since it is no longer applied to `sequence`.

File: torch_frame/data/dataset.py [hunk 7]
```diff
@@ -262,7 +266,7 @@ def __init__(
         self.text_embedder_cfg = text_embedder_cfg
         self.col_to_sep = canonicalize_col_to_sep(col_to_sep, [
             col for col, stype in self.col_to_stype.items()
-            if stype == torch_frame.multicategorical
+            if stype in [torch_frame.multicategorical, torch_frame.sequence]
```
Comment (weihua916) [inline torch_frame/data/dataset.py:269]: revert

File: torch_frame/data/mapper.py [hunk 8]
```diff
@@ -165,6 +165,47 @@ def backward(self, tensor: MultiNestedTensor) -> pd.Series:
         return pd.Series(ser)
 
 
+class SequenceTensorMapper(TensorMapper):
+    r"""Maps any sequence series into an :obj:`MultiNestedTensor`.
+
+    Args:
+        sep (str): The delimiter for the sequence in each cell.
+        (default: :obj:`,`)
```
Comment (weihua916) [inline torch_frame/data/mapper.py:173]: revert

File: torch_frame/data/mapper.py [hunk 9]
```diff
@@ -165,6 +165,47 @@ def backward(self, tensor: MultiNestedTensor) -> pd.Series:
...
+        values = []
+        num_rows = len(ser)
+        offset = ser.apply(lambda row: len(row) if row is not None and not (
+            isinstance(row, float) and pd.isna(row)) else 0)
+        ser = ser[offset != 0]
+        offset = pd.concat((pd.Series([0]), offset))
+        offset = torch.from_numpy(offset.values)
+        offset = torch.cumsum(offset, dim=0)
+        ser = ser.explode()
+        ser = ser.astype('float64')
```
Comment (weihua916) [inline torch_frame/data/mapper.py:193]: float32

File: torch_frame/data/mapper.py [hunk 10]
```diff
@@ -165,6 +165,47 @@ def backward(self, tensor: MultiNestedTensor) -> pd.Series:
...
+    def forward(
+        self,
+        ser: Series,
+        *,
+        device: Optional[torch.device] = None,
+    ) -> MultiNestedTensor:
+        values = []
+        num_rows = len(ser)
+        offset = ser.apply(lambda row: len(row) if row is not None and not (
+            isinstance(row, float) and pd.isna(row)) else 0)
```
Comment (weihua916) [inline torch_frame/data/mapper.py:187]: I think it is safer to check if each element of `ser` is a list or not.

File: torch_frame/data/stats.py [hunk 11]
```diff
@@ -30,24 +30,16 @@ class StatType(Enum):
...
+            ],
+            torch_frame.categorical: [StatType.COUNT],
+            torch_frame.multicategorical: [StatType.MULTI_COUNT]
+        }
+        return stats_type.get(stype, [])
```
Comment (weihua916) [inline torch_frame/data/stats.py:42]: Nice!

File: torch_frame/data/mapper.py (hunk 10 above)
Comment (yiweny) [reply torch_frame/data/mapper.py]: Fixed

File: torch_frame/data/mapper.py [hunk 12]
```diff
@@ -165,6 +165,51 @@ def backward(self, tensor: MultiNestedTensor) -> pd.Series:
...
+        values = []
+        num_rows = len(ser)
+        offset = ser.apply(lambda row: self.get_sequence_length(row))
+        ser = ser[offset != 0]
+        offset = pd.concat((pd.Series([0]), offset))
+        offset = torch.from_numpy(offset.values)
+        offset = torch.cumsum(offset, dim=0)
+        ser = ser.explode()
+        ser = ser.astype('float32')
+        values = torch.from_numpy(ser.values)
```
Comment (weihua916) [inline torch_frame/data/mapper.py:198]: nit: 
```python
values = torch.from_numpy(ser.values.astype('float32'))
```

File: torch_frame/data/mapper.py [hunk 13]
```diff
@@ -165,6 +165,51 @@ def backward(self, tensor: MultiNestedTensor) -> pd.Series:
...
+        return MultiNestedTensor(num_rows=num_rows, num_cols=1, values=values,
+                                 offset=offset)
+
+    def backward(self, tensor: MultiNestedTensor) -> pd.Series:
+        values = tensor.values
+        offset = tensor.offset
+        values = values.tolist()
```
Comment (weihua916) [inline torch_frame/data/mapper.py:205]:
```python
values = tensor.values.cpu().numpy()
```
should be good

**Review by weihua916 — APPROVED** [review]

**Pull request merged**